src/charm.py

```diff
@@ -0,0 +1,661 @@
+        repositories = list()
+        if cm_access_rules := self.access_rules_configmap.get():
+            for key in cm_access_rules.keys():
+                repositories.append(f"{self._access_rules_dir_path}/{key}")
+        return repositories
+
+    def _render_conf_file(self) -> str:
+        """Render the Oathkeeper configuration file."""
+        with open("templates/oathkeeper.yaml.j2", "r") as file:
+            template = Template(file.read())
+
+        rendered = template.render(
+            kratos_session_url=self._kratos_session_url,
+            kratos_login_url=self._kratos_login_url,
+            access_rules=self._get_all_access_rules_repositories(),
+        )
+        return rendered
+
+    @retry(
+        wait=wait_exponential(multiplier=3, min=1, max=10),
+        stop=stop_after_attempt(5),
+        reraise=True,
+        before=before_log(logger, logging.DEBUG),
+    )
+    def _update_config(self) -> None:
+        conf = self._render_conf_file()
+        self.oathkeeper_configmap.update({"oathkeeper.yaml": conf})
+
+        if all(
+            [
+                self.cert_handler.cert,
+                self.cert_handler.key,
+                self.cert_handler.ca,
+            ]
+        ):
+            self.update_cert_configuration(
+                self.cert_handler.cert, self.cert_handler.key, self.cert_handler.ca
+            )
+
+    def _get_kratos_endpoint_info(self, key: str) -> Optional[str]:
+        if not self.model.relations[self._kratos_relation_name]:
+            logger.info("Kratos relation not found")
+            return
+
+        try:
+            kratos_endpoints = self.kratos_endpoints.get_kratos_endpoints()
+            return kratos_endpoints[key]
+        except KratosEndpointsRelationDataMissingError:
+            logger.info("No kratos-endpoint-info relation data found")
+            return
+
+    def _auth_proxy_relation_peer_data_key(self, relation_id: int) -> str:
+        return f"auth_proxy_{relation_id}"
+
+    @property
+    def _peers(self) -> Optional[Relation]:
+        """Fetch the peer relation."""
+        return self.model.get_relation(PEER)
+
+    def _set_peer_data(self, key: str, data: Dict) -> None:
+        """Put information into the peer data bucket."""
+        if not (peers := self._peers):
+            return
+        peers.data[self.app][key] = json.dumps(data)
+
+    def _get_peer_data(self, key: str) -> Dict:
+        """Retrieve information from the peer data bucket."""
+        if not (peers := self._peers):
+            return {}
+        data = peers.data[self.app].get(key, "")
+        return json.loads(data) if data else {}
+
+    def _pop_peer_data(self, key: str) -> Dict:
+        """Retrieve and remove information from the peer data bucket."""
+        if not (peers := self._peers):
+            return {}
+        data = peers.data[self.app].pop(key, "")
+        return json.loads(data) if data else {}
```

Comment:
That's true, but I see no problem with doing it. Sometimes when you access the same object attribute more than once, you can assign it to a variable to avoid having to access the object every time